pydantic/_internal/_decorators.py

```diff
@@ -717,37 +717,33 @@ def unwrap_wrapped_function(
     unwrap_class_static_method: bool = True,
 ) -> Any:
     """Recursively unwraps a wrapped function until the underlying function is reached.
-    This handles property, functools.partial, functools.partialmethod, staticmethod and classmethod.
+    This handles property, functools.partial, functools.partialmethod, staticmethod, and classmethod.
 
     Args:
         func: The function to unwrap.
-        unwrap_partial: If True (default), unwrap partial and partialmethod decorators, otherwise don't.
-            decorators.
+        unwrap_partial: If True (default), unwrap partial and partialmethod decorators.
         unwrap_class_static_method: If True (default), also unwrap classmethod and staticmethod
             decorators. If False, only unwrap partial and partialmethod decorators.
 
     Returns:
         The underlying function of the wrapped function.
     """
-    all: set[Any] = {property, cached_property}
-
-    if unwrap_partial:
-        all.update({partial, partialmethod})
-
-    if unwrap_class_static_method:
-        all.update({staticmethod, classmethod})
+    # Define the types we want to check against as a single tuple.
+    unwrap_types = (
+        (property, cached_property)
+        + (partial, partialmethod if unwrap_partial else ())
+        + (staticmethod, classmethod if unwrap_class_static_method else ())
+    )
```

Comment:
```suggestion
    unwrap_types = (
        (property, cached_property)
        + (partial, partialmethod) if unwrap_partial else ()
        + (staticmethod, classmethod) if unwrap_class_static_method else ()
    )
```

Reply:
thank you for this, I played a bit with the different ways of doing this and found the best one which had the correct behavior. Updated the PR with the change